### ISSUE TITLE
Check doc comment only for public methods.

### ISSUE THREAD

**Issue by obstschale** [squizlabs/PHP_CodeSniffer#2077]
I am aware of the `Squiz.Commenting.BlockComment` sniff. But is it possible to only check BlockComments for `public` methods and ignore `private` and `protected` ones?

**Comment by gsherwood** [comment]
There is no option to do that with the current sniff. A sniff property would need to be created for this.

**Comment by obstschale** [comment]
Unfortunately, I don't know how to do this. So I have to wait for you or someone to implement this. Thank you for your reply.

**Comment by VincentLanglet** [comment]
> Unfortunately, I don't know how to do this. So I have to wait for you or someone to implement this. Thank you for your reply.

Are you still interested ?

**Comment by obstschale** [comment]
Yes, I am. But still, I don't know how and I currently have no time to dig into it. :)

**Comment by VincentLanglet** [comment]
I can make a try. What is you exact need ?
I have difficulty to understand how/why you want to check only `public` method.

For what I understand from this sniff this is how it works.
```
/** This comment is checked */

/**
 * This comment is not checked
 *
 * @param $a
 */
function foo() {
 /** This comment is checked */
}
```

**Comment by obstschale** [comment]
We have decided at our team to document every public method, cause they act like an public API for other parts of our code / modules. So these methods should be documented. private and protected methods are not mandatory to document, cause they are only used withing that class.

Example:

```php
class Foo {

  /**
   * Mandatory DocBlock.
   * If this is DocBlock is missing
   * CodeSniffer should fail.
   */
  public function one() {}

  /**
   * Not mandatory.
   * Code Sniffer does not care if this exists.
   */
  protected function two() {}


  /**
   * Not mandatory.
   * Code Sniffer does not care if this exists.
   */
  private function two() {}

}
```

**Comment by VincentLanglet** [comment]
@obstschale Ok, I understand.
But you talked about the `Squiz.Commenting.BlockComment` sniff, but I think this sniff does not check there is a `BlockComment function`.

For what I know, this is the `Squiz.Commenting.FunctionComment` sniff, extending the `PEAR.Commenting.FunctionComment` one.

These sniffs are checking :
- Function has doc comment
- `/**` style is used for function comment
- The @return tag
- The @param tags
- The @throws tags.

And then if you also use the `Generic.Commenting.DocComment` sniff, it checks the doc comment will start with a short description.

So it seems more difficult to make something coherent, I need more example.

```
class Foo {

  /**
   * You say it's mandatory, even if there is no need for param/return/throws tag
   */
  public function one() {}

  /**
   * You say it's not mandatory
   */
  protected function two() {}

  /**
   * It is still not mandatory for you even if it needs a param/return tag ?
   */
  private function three(int $foo) { return 1; }
}
```

Personally, in my project, if the function needs no `@param/@return/@throws` tag, the phpDoc is not needed (for protected, private but also public function). I can understand someone who want this rule only for protected/private ones, or only private ones.

If the expected rule is "The phpDoc is not needed for protected/private function even if they need @param/@return/@throws tags", it's easy to add an option, but I feel like it's very specific and then maybe it will need more feedback, because I'm not sure @gsherwood will accept an option used by a single person/team/project.

**Comment by obstschale** [comment]
It seems I didn't understand the difference between `Squiz.Commenting.FunctionComment` and `Squiz.Commenting.BlockComment` when I opened that issue.

Your explaination is good. I want `Squiz.Commenting.FunctionComment` to be enhanced. You assumption was correct. The only option I would need is to enabled or disabled the rule for certain function types. So even if a `private` functions would need `@param/@return/@throws` comments, but the config disabled the rule for `private` functions, CodeSniffer would not thrown en error (maybe an warning, if it were possible).

**Comment by VincentLanglet** [comment]
Ok, I think this would be easy.

**Comment by VincentLanglet** [comment]
A PR is opened @obstschale
https://github.com/squizlabs/PHP_CodeSniffer/pull/2815

**Comment by gsherwood** [comment]
Thanks a lot @VincentLanglet for the PR. This new feature is now merged in and will be in the 3.6.0 release.

Documentation for the new sniff property is here: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties#pearcommentingfunctioncomment
The Squiz sniff also inherits this property, and is documented here: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties#squizcommentingfunctioncomment
